name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TastytradePDFExtractor.java

```diff
@@ -0,0 +1,529 @@
+                t.setCurrencyCode("USD");
+                v.put("currency", "USD");
+                t.setSecurity(getOrCreateSecurity(v));
+                t.setDate(asDate(v.get("tradedate")));
+                t.setShares(asShares(v.get("shares"), "en", "US"));
+                t.setAmount(asAmount(v.get("amount"), "en", "US"));
+                checkAndSetFee(calculateFee(v), t, type.getCurrentContext());
+            }))
+
+            .wrap(BuySellEntryItem::new));
+
+        // Expired Option
+        //    EXPIRED 04/06/23 M PUT  TQQQ   04/06/23    27 1  
+        //    PROSHARES ULTRAPRO QQQ
+        //    OPTION EXPIRATION - EXPIRED
+        //    Security Number: 9LVCJB6
+        // Option Assignment
+        //    ASG 06/09/23 C CALL UPRO   06/09/23    40 1  
+        //    PROSHARES ULTRAPRO S&P 500
+        //    A/E 9SMCXK2 1 ASSIGNED
+        //    Security Number: 9SMCXK2
+        Block optionExpiration = new Block("^(EXPIRED|ASG).*$");
+        type.addBlock(optionExpiration);
+        optionExpiration.set(new Transaction<BuySellEntry>().subject(() -> {
+            BuySellEntry entry = new BuySellEntry();
+            entry.setMonetaryAmount(Money.of("USD", 0));
+            return entry;
+        })        
+            .section("date", "type",  "otype", "underlaying", "expiration", "strike", "quantity", "description", "cusip")
+            .match("^(?<type>(EXPIRED|ASG))[\\s]+(?<date>[\\d]{2}/[\\d]{2}/[\\d]{2})[\\s]+[A-Z][\\s]+(?<otype>[PUTCAL]{3,4})[\\s]+"
+                            + "(?<underlaying>[/\\&\\w]+)[\\s]+"
+                            + "(?<expiration>[\\d]{2}/[\\d]{2}/[\\d]{2})[\\s]+"
+                            + "(?<strike>[,\\.\\d]+)[\\s]+"
+                            + "(?<quantity>[\\-]?[\\d]+)[\\s]*.*$")
+            .match("^(?<description>.*).*$")
+            .match("(^OPTION EXPIRATION.*$|^.*[\\s]+[A-Z0-9]+[\\s]+[\\d]+[\\s]+ASSIGNED.*)")
+            .match("Security Number:[\\s]+(?<cusip>[A-Z0-9]+)")
+            .assign((t, v) -> {
                v.put("currency", "USD");
+                v.put("wkn", v.get("cusip"));
+                t.setDate(asDate(v.get("date")));
+                OccOsiSymbology o = new OccOsiSymbology(v.get("underlaying"),
+                                v.get("expiration"), v.get("type"),
+                                Double.parseDouble(v.get("strike").toString()));
+                v.put("tickerSymbol", o.getOccKey());
+                v.put("name", o.getName());
+                t.setShares(asShares(v.get("quantity").replace("-", ""), "en", "US") * 100);
+                t.setSecurity(getOrCreateSecurity(v));
+                if (v.get("type").equals("EXPIRED")) {
+                    setTypeAndExpirationNote(t, v); // TODO
+                } else {
+                    t.setNote("Assigned: Buy To Close");
+                    t.setType(PortfolioTransaction.Type.BUY); // Assignment means this was a short position
+                }
```

Comment:
In this function you already set the type "Buy" above.
So if the type should change, then you could, if you want, use the following code

Maybe it's better to use a separate section like
```
                .section("type").optional()
                .match("^XXXXX (?<type>YYYYYYY)$")
                .assign((t, v) -> {
                    if ("YYYYYYY".equals(v.get("type")))
                        t.setType(PortfolioTransaction.Type.SELL);
                })
```

Reply:
Depending on the type (expiration or assignment), different transactions are possible.
- Long and short positions can expire "worthless" out of the money. That means whenever we see an expiration transaction in the Account Statement, we need to check whether it was a long or a short position at expiration.
- Only short positions can be assigned, which happens when the contract expires in the money. So we do not need to check the quantity.

Here is a hopefully more readable and understandable version.
```suggestion
                // Expiration means it can be a long or short position that expired OTM
                if (v.get("type").equals("EXPIRED")) {
                    BigDecimal quantity = ExtractorUtils.convertToNumberBigDecimal(v.get("quantity"), Values.Amount, "en", "US");
                    if (quantity.signum() == 1) {
                        t.setType(PortfolioTransaction.Type.BUY);
                        t.setNote("Expired: Buy To Close");
                    }
                    else if (quantity.signum() == -1) {
                        t.setType(PortfolioTransaction.Type.SELL);
                        t.setNote("Expired: Sell To Close");
                    }
                // Assignment means this was a short position
                } else if (v.get("type").equals("ASG")) {
                    t.setNote("Assigned: Buy To Close");
                    t.setType(PortfolioTransaction.Type.BUY);
                }
```